### PR TITLE
More tests and position calculations

### PR DESCRIPTION
**Pull request by GLRoylance**
## Overview

This PR tests positions of the field elements and other calculations. A disabled test shows Blue Alliance poses facing wrong direction. Calculations show Game Manual misleads with statement that bottom of AprilTags are so high off the carpet; the real description should be the bottom AprilTag mounting screws are so high off the carpet.

## Design Doc

Game Manual specs; Field Element documents.

## Tests Ran

It is a set of unit tests. Failing tests disabled.

### CODE REVIEW

File: src/test/java/frc/robot/util/FieldElementTest.java [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+    double outerGrid = 6.0 * 12 + 3;
+    // The inner Grid is 5 feet 6 inches (Game Manual)
+    double innerGrid = 5.0 * 12 + 6;
+
+    // Outermost hybrid nodes are 2 feet 1.75 inches wide (Game Manual)
+    double outerHybrid = 2.0 * 12 + 1.75;
+    // other hybrid nodes are 1 feet 6.5 inches wide (Game Manual)
+    double innerHybrid = 1.0 * 12 + 6.5;
+    // width of bar (consistent value chosen from Game Manual dimensions)
+    double bar = 3.5;
```
Comment (AnthonyFurman) [inline src/test/java/frc/robot/util/FieldElementTest.java:48]: what do you mean by bar?

File: src/test/java/frc/robot/util/FieldElementTest.java [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+        Node n = new Node((Vision)null, Alliance.Red, 2, 5);
+        double ang = n.scorePose.getRotation().getRadians();
+        ang = Math.IEEEremainder(ang, 2 * Math.PI);
+        assertEquals(0.0, ang, 0.000001);
+    }
+
+    /**
+     * Transform a Blue Alliance pose to a Red Alliance pose.
+     * 
+     * <p> This should be a library method.
```
Comment (AnthonyFurman) [inline src/test/java/frc/robot/util/FieldElementTest.java:185]: It should be. Is there a reason it is here?

File: src/test/java/frc/robot/util/FieldElementTest.java [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+        // calculate the expected pose
+        Pose2d poseExpect = new Pose2d(widthField - 2.1, 3.6, new Rotation2d(0.75 * Math.PI));
+ 
+        // broken out tests
+        assertEquals(poseExpect.getX(), poseCheck.getX(), 0.001);
+        assertEquals(poseExpect.getY(), poseCheck.getY(), 0.001);
+        assertEquals(poseExpect.getRotation().getRadians(), poseCheck.getRotation().getRadians(), 0.001);
+
+        // integrated test
+        assertTrue(testEquals(poseCheck, poseExpect, 0.001));
```
Comment (AnthonyFurman) [inline src/test/java/frc/robot/util/FieldElementTest.java:253]: is there a reason to do both integrated and seperate tests?

**Comment by AnthonyFurman** [comment]
I want someone, @PandaNinja10 or at @omegaladon (I know both of you worked with the node class) to check why some of the values are wrong, and to potentially add staged game pieces. That can be a seperate PR but I want them to look at this.

File: src/main/java/frc/robot/constants/FieldConstants.java [hunk 4]
```diff
@@ -11,19 +11,10 @@ public class FieldConstants {
   // TODO: After testing, this value might need to be tuned. Change the 0+ to however much the robot needs to be away from the node tape
   public static final double kRobotDistanceFromNodeTape = 0 + DriveConstants.kRobotWidthWithBumpers / 2; // meters
 
-  // The locations of the node tape in the x axis, relative to the very bottom left corner of the entire field
-  public static final double kBlueNodeTapePosX = 4.379; // meters
-  public static final double kRedNodeTapePosX = 18.183; // meters
+  // The distance from the apriltag to the edge of the tape in front of the grids
+  public static final double kAprilTagOffset = Units.inchesToMeters(14.06 - 0.26); // meters
```
Comment (richtan) [inline src/main/java/frc/robot/constants/FieldConstants.java:15]: So this is 14.25 inches (from the surface of the grid that the AprilTag is on to the end of the hybrid node divider) - 0.22 in (thickness of AprilTag plate) = 14.03 inches?

File: src/main/java/frc/robot/constants/FieldConstants.java [hunk 5]
```diff
@@ -11,19 +11,10 @@ public class FieldConstants {
...
-  public static final double kBlueAllianceNodeStartX = kBlueNodeTapePosX + kRobotDistanceFromNodeTape; // meters
-  // Subtract because the red alliance is on the opposite side of the field
-  public static final double kRedAllianceNodeStartX = kRedNodeTapePosX - kRobotDistanceFromNodeTape; // meters
-
-  // The location of the field boundary in the y axis, relative to the very bottom left corner of the entire field
-  public static final double kNodeStartY = 0.569; // meters
-
-  // TODO: This value needs to be tested to ensure functionality, for node scoring
-  public static final double kRobotOffsetX = DriveConstants.kRobotWidthWithBumpers / 2; // meters
+  // The distance from center of apriltag to the cone node
```
Comment (richtan) [inline src/main/java/frc/robot/constants/FieldConstants.java:17]: What distance is this referring to?

File: src/main/java/frc/robot/controls/Operator.java [hunk 6]
```diff
@@ -110,7 +110,7 @@ public void selectValue(NodePositionIndex index, int value) {
     if (value == 0 && DriverStation.getAlliance() == Alliance.Red) value = 2;
```
Comment (richtan) [inline src/main/java/frc/robot/controls/Operator.java:110]: Won't this switch the row as well?

File: src/main/java/frc/robot/util/Node.java [hunk 7]
```diff
@@ -54,48 +54,38 @@ public Node(Vision vision, Alliance alliance, int row, int column) {
           : NodeType.CONE;
 
     // Starting locations
-    double x = (alliance == Alliance.Blue ? 
-      FieldConstants.kBlueAllianceNodeStartX : 
-      FieldConstants.kRedAllianceNodeStartX
-    );
+    double x = 0;
+    if (alliance == Alliance.Blue) {
+      x = VisionConstants.kAprilTags.get(5).pose.getX() + FieldConstants.kAprilTagOffset + FieldConstants.kRobotDistanceFromNodeTape;
```
Comment (richtan) [inline src/main/java/frc/robot/util/Node.java:59]: AprilTag 5 is on the double substation, I assume you want an AprilTag on the grid

Comment (richtan) [inline src/main/java/frc/robot/util/Node.java:59]: Robot width is 26 inches, bumpers are 3 inches thick on each side, and there is about a 0.25 inch gap between the bumper and the chassis, so the robot width with bumpers on should be 32.5 inches total, not 33.5 inches.

Comment (omegaladon) [reply src/main/java/frc/robot/util/Node.java:59]: `kRobotWidthWithBumpers = Units.inchesToMeters(26 + 3.75 * 2);` should already be 33.5 inches

File: src/test/java/frc/robot/util/FieldElementTest.java (hunk 3 above)
Comment (AnthonyFurman) [reply src/test/java/frc/robot/util/FieldElementTest.java:266]: Writing tests of unit tests would be beyond redundant.
As for the point about unit tests, I found they increase my builds by ~2-4 seconds.
Lets be safe and say 0.1 second.
0.1 billion seconds = 3.1 years.
Of course, that's all of the unit tests together currently on main, and it's probably some initialization that takes that time.
I don't really see a justification for or against the test.

**Review by AnthonyFurman — APPROVED** [review]

File: src/test/java/frc/robot/util/FieldElementTest.java [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+    @AfterEach
+    public void cleanup() {}
+
+    /**
+     * Check field dimensions
+     */
+    @Test
+    public void testFieldDimensions() {
+        // TODO: terminology: kFieldWidth is confusing because field is drawn landscape, so width != length
+        assertEquals(heightField, FieldConstants.kFieldWidth, 0.001);
```
Comment (vfurman-gh) [inline src/test/java/frc/robot/util/FieldElementTest.java:63]: Here and throughout this PR: I do not understand why do we need tests that assert that constants have certain values. Is this to make harder to change constants? This seems counter-productive.

**Pull request merged**